test/samplers/delayed_sampling.js

```diff
@@ -0,0 +1,223 @@
+      this._matchers = {};
+      matchers.forEach(m => {
+        this._matchers[m.tagValue] = m;
+      });
+      this._undecided = { sample: false, retryable: true, tags: null };
+    }
+
+    _findTag(tags: Array<Tag>): ?Tag {
+      for (let i = 0; i < tags.length; i++) {
+        if (tags[i].key === this._tagKey) {
+          return tags[i];
+        }
+      }
+      return null;
+    }
+
+    _createOutTags(tagValue: string): { [string]: string } {
+      return {
+        'sampler.type': 'TagEqualsSampler',
+        'sampler.param': tagValue,
+      };
+    }
+
+    _decide(tagValue: any): SamplingDecision {
+      const match: ?Matcher = this._matchers[tagValue];
+      if (match) {
+        return { sample: true, retryable: false, tags: this._createOutTags(match.tagValue) };
+      }
+      return this._undecided;
+    }
+
+    onCreateSpan(span: Span): SamplingDecision {
+      const tag: ?Tag = this._findTag(span.getTags());
+      if (tag) {
+        return this._decide(tag.value);
+      }
+      return this._undecided;
+    }
+
+    onSetOperationName(span: Span, operationName: string): SamplingDecision {
+      return this.onCreateSpan(span);
+    }
+
+    onSetTag(span: Span, key: string, value: any): SamplingDecision {
+      if (key === this._tagKey) {
+        return this._decide(value);
+      }
+      return this._undecided;
+    }
+
+    toString(): string {
+      return 'TagEqualsSampler';
+    }
+
+    close(callback: ?Function): void {
+      if (callback) {
+        callback();
+      }
+    }
+  }
+
+  declare type PrioritySamplerState = {
+    samplerFired: Array<boolean>,
+  };
+
+  /**
+   * PrioritySampler contains a list of samplers that it interrogates in order.
+   * Sampling methods return as soon as one of the samplers returns sample=true.
+   * The retryable state for each underlying sampler is stored in the extended context
+   * and once retryable=false is returned by one of the delegates it will never be
+   * called against.
+   *
+   * TODO: since sampling state is shared across all spans of the trace, the execution
+   *       of the sampler should probably only happen on the local-root span.
+   */
+  class PrioritySampler implements Sampler {
+    apiVersion = SAMPLER_API_V2;
+
+    _delegates: Array<Sampler>;
+
+    constructor(samplers: Array<Sampler | LegacySamplerV1>) {
+      this._delegates = samplers.map(s => adaptSamplerOrThrow(s));
+    }
+
+    _getState(span: Span): PrioritySamplerState {
+      const store = span.context()._samplingState.extendedState();
+      const stateKey = 'PrioritySampler'; // TODO ideally should be uniqueName() per BaseSamplerB2
+      let state: ?PrioritySamplerState = store[stateKey];
+      if (!state) {
+        state = {
+          samplerFired: Array(this._delegates.length),
+        };
+        // cannot use array.fill() in Node 0.10
+        for (let i = 0; i < this._delegates.length; i++) {
+          state.samplerFired[i] = false;
```

Comment:
Can't this be encapsulated within `PrioritySamplerState`

Reply:
+1